### DISCUSSION TITLE
more powerful apply rule and future of apply

### DISCUSSION THREAD

**Discussion by minimit** [tailwindlabs/tailwindcss#2106] [Ideas]
I've studied preprocessors mixins and extends: all the preprocessor have different way to use mixins and extend rules.

Scss/Sylius/PostcssMixins I don't really like that you have to specify the mixins before using it, so you can't really make mixins as you go and use them, you end up instead doing like Bootstrap does and include all mixins as first thing for complex things, instead of using mixins for normal styles.

PostcssExtendRule have some problems (!important doesn't work, inside media queries doesn't work) and it's a bit stale (many postcss plugins are I don't know why and it really discourages me).

ScssExtend/PostcssExtend I find them really screw up the css specificity, if you extend something you end up with different specificity because they add the selector to the extended one.

And finally there's tailwind with `apply`, I like tailwind because it has an `@apply` rule, but I'm also eager to know what's the future of apply, will it be like less mixins (that use all selector also if nested, they are very powerful but also dangerous for file size) or will it be removed with no substitute solution.

Here's what @adamwathan told us about the future of apply:
https://github.com/tailwindlabs/tailwindcss/issues/1599#issuecomment-618531722
https://github.com/tailwindlabs/tailwindcss/issues/1897#issuecomment-650610098

I'd really like a powerful apply that apply all the selectors like this (like less does), since you need to add the `.animate__infinite` class if you want that also.

```
.animate__animated {
  -webkit-animation-duration: 1s;
  animation-duration: 1s;
  -webkit-animation-duration: var(--animate-duration);
  animation-duration: var(--animate-duration);
  -webkit-animation-fill-mode: both;
  animation-fill-mode: both;
}
.animate__animated.animate__infinite {
  -webkit-animation-iteration-count: infinite;
  animation-iteration-count: infinite;
}

.foo {
  @apply animate__animated;
}

/* result */

.foo {
  -webkit-animation-duration: 1s;
  animation-duration: 1s;
  -webkit-animation-duration: var(--animate-duration);
  animation-duration: var(--animate-duration);
  -webkit-animation-fill-mode: both;
  animation-fill-mode: both;
}
.foo.animate__infinite {
  -webkit-animation-iteration-count: infinite;
  animation-iteration-count: infinite;
}
```

Or at least I'd like an apply that works with pseudo selectors that right now really limit the apply usage for normal single css components.

And last but not least, the proposals of a css extend/apply rule are failed and right now there's no future roadmap for a css css extend/apply rule, just a solution with Shadow DOM I don't really understand (never used shadow dom).
https://github.com/w3c/csswg-drafts/issues/1047
https://discourse.wicg.io/t/needed-new-champion-for-css-apply-rule/2012/19

**Comment by minimit** [comment]
solved with this #2159

**Reply by neoswf** [reply]
Thank you man! saved me lots of time, to find this solution. Thank you!!!
